### ISSUE TITLE
AttributeError: 'DataFrame' object has no attribute 'name'; Various stack overflow / github suggested fixes not working

### ISSUE THREAD

**Issue by david-thrower** [dask/dask#8624]
**What happened**:

I perform a pipeline of transformations on a Dask dataframe originating from dd.read_sql_table() from a view in an oracle DB. In one stage that follows many successful stages, I try to apply a simple class method to a column of type <object> (strings). This always returns this error (main issue):

```
{

File "[redacted]/pandas/core/generic.py", line 5487, in __getattr__
    return object.__getattribute__(self, name)
AttributeError: 'DataFrame' object has no attribute 'name'. Did you mean: 'rename'?

}
```
When I tried to inspect the dataframe object having this column on which the .apply() was not working, to see if there was a defect in this dataframe object as an artifact of upstream operations, I get this error (secondary issue):

```
{
Traceback (most recent call last):
  ...
  File "[redacted]/python3.10/inspect.py", line 469, in getmembers
    value = getattr(object, key)
  File "[redacted]/dask/dataframe/core.py", line 4102, in empty
    raise NotImplementedError(
NotImplementedError: Checking whether a Dask DataFrame has any rows may be expensive. However, checking the number of columns is fast. Depending on which of these results you need, use either `len(df.index) == 0` or `len(df.columns) == 0`
}
```

**What you expected to happen**:

1. The method being applied should return a Dask dataframe of 3 np.int64 columns:  state_id, city_id,  district_id.
2. inspect.getmembers(my_dask_df) was expected to return a list of object names and values to introspect the dask dataframe object.

**Minimal Complete Verifiable Example**:

1. Starting with a Dask dataframe returned by a chain of upstream sorting and filtering operations (df.query(), boolean masking, etc ...) on an original dataframe returned by dd.read_sql_table() from an Oracle DB which looks like this:

```
{

#       house_id             city_and_state                     district                                     ...    has_leaky_roof           owes_taxes
#       <object>                   <object>                        <object>                                              <object>                 <object>
#     1    "1"                      "Tempe, AZ"            "Tempe, AZ, University"                                "no"                             "no"
#     2   "17"                  "Santa Monica, CA"    'Santa Monica, CA N, Ocean'                       "no"                             "no"

}
```

2. I want to take the column "district" and call its Series.apply() method to return a dataframe that looks like this:

```
            state_id         city_id        district_id
           <int64>         <int64>        <int64>
                    2               8675309      5551212
                    5               1234                5678

```
To do this, I add a method to a class. The method takes a string as an argument, returns a Dask series representing the state_id,city_id, and district_id found when querying that district (returns a row of the intended resultant dataframe).   
  
```python3

# in a package ... module: myimportutils.my_import_utils.py:


class SomeUtils:
    """ Some docstring ... """
    def __init__(self, dd, pd, homes_to_import_unnormalized_dd_df ...)
        self.dd = dd
        self.pd = pd
        self.homes_to_import_unnormalized_dd_df =\
            homes_to_import_unnormalized_dd_df
        ...

    # Create a pandas DataFrame having the information needed to look up 
    # the state_id, city_id, district_id, given the name of a district:

    self.pd_df_of_ids = self.pd.read_csv('distinct_cities_to_city_and_state_ids.csv')

    # self.pd_df_of_id looks like this:
 
    # district                                   state_id                city_id               district_id
    # <object>                            <numpy.int8>     <numpy.int64>    <numpy.int64>
    # "Tempe, AZ, Unicersity"               2                  8675309              5551212
    # 'Santa Monica, CA N, Ocean '     5                    1234                     5678
    # ...

    # Define a method that takes a district as a string an returns the Dask series
    def find_city_and_state(self,district_str)
        query = "district == '{district_str}'"
        row = self.pd_df_of_ids.query(query)
        if row.shape[0] == 1:
            locale_info_pd_Series =\
                pd.series([
                           row['state_id'].values[0],
                           row['city_id'].values[0],
                           row['district_id'].values[0]
                           ],
                          index=[
                                 'state_id',
                                 'city_id',
                                 'district_id'])
        else:
            locale_info_pd_series =\
                pd.Series([
                           0.
                           0,
                           0
                           ],
                          index=[
                                 'state_id',
                                 'city_id',
                                 'district_id'])
        locale_info_dd_series= self.dd.from_pandas(locale_info_pd_series,
                                                        chunksize=800)
        return  locale_info_dd_series

    # Define a method that applies the last method to the column "district":  
    def get_df_of_ids(self):
        locale_info_df = self.homes_to_import_unnormalized_dd_df["district"].apply(
            self.find_city_and_state,
            meta=self.dd.utils.make_meta([('state_id', np.int64),
                                          ('city_id', np.int64),
                                          ('district_id', np.int64)])).compute()
        return locale_info_dd_df

}
```
Then in the main script where I had done all the upstream transformations to the dataframe which we are working on, instantiate this class and call its get_df_of_ids method:

```python3
{

from myimportutils.my_import_utils import SomeUtils

tools = SomeUtils(homes_to_import_unnormalized_dd_df)
df_of_locale_ids = tools.get_df_of_ids() # This line fails:

}
```

This is the Error that this is giving me:

```
{
File "[
... 
redacted]/pandas/core/generic.py", line 5487, in __getattr__
    return object.__getattribute__(self, name)
AttributeError: 'DataFrame' object has no attribute 'name'. Did you mean: 'rename'?
}
```
If I do this:

```python3
{

# in myimportutils.my_import_utils.py:

from inspect import getmembers

class SomeUtils:
    """ Some docstring ... """
    def __init__(self, dd, pd, homes_to_import_unnormalized_dd_df ...)
        self.dd = dd
        self.pd = pd
        self.homes_to_import_unnormalized_dd_df =\
            homes_to_import_unnormalized_dd_df
        ...

    def test(self)
        print(getmembers(self.homes_to_import_unnormalized_dd_df))

 }
```
In my main script:

```python3
{

from myimportutils.my_import_utils import SomeUtils

tools = SomeUtils(homes_to_import_unnormalized_dd_df)
tools.test()
}
```
I get this error:

```
{

Traceback (most recent call last):
  ...
  File "[redacted]/python3.10/inspect.py", line 469, in getmembers
    value = getattr(object, key)
  File "[redacted]/dask/dataframe/core.py", line 4102, in empty
    raise NotImplementedError(
NotImplementedError: Checking whether a Dask DataFrame has any rows may be expensive. However, checking the number of columns is fast. Depending on which of these results you need, use either `len(df.index) == 0` or `len(df.columns) == 0`

}
```

**Environment**:

- Dask version:  2021.7.0
- Python version: 3.10.0
- Operating System: Alpine container with Python and Dask dependencies.
- Install method: pip

**Comment by ncclementi** [comment]
@david-thrower thank you for the detailed explanation, is there a possibility that you can provide a more concise minimal reproducible example that we can try on our end. It's a little difficult to help you. In the meantime I noticed couple of things that might help:

- There are couple of related issue raised on pandas and sklearn that might have some ideas to try, and it looks like in most cases the problem comes at some point where you are probably passing a data frame where  you should be passing a Series object
https://github.com/twopirllc/pandas-ta/issues/226
https://github.com/scikit-learn-contrib/imbalanced-learn/issues/666
- You are using a very old version of dask and maybe pandas. It might be worth updating to a newer one.
- I don't think python 3.10 is the problem but keep in mind that python 3.10 is not "supported" yet, meaning tested on CI.

**Comment by david-thrower** [comment]
Sorry for the longer explanation. I included the entire pipeline because on the original post, because the issue appears unusual and may be an artifact of something upstream from the actual affected step. I summarized it to the extent possible and fixed a few typos.

**Comment by david-thrower** [comment]
@jsignell, @bryanwweber, @ncclementi  Thank you all for the replies and sorry for the broken communication on this complex and cryptic issue. I also appreciate your patience as  struggle to articulate the history of troubleshooting I have done and  not be too verbose.  

So for clarification, for the suggestion of upgrading Pandas and the possibility that the issue may have been caused by using an obsolete / deprecated version of Pandas, it appears I was on a version of Pandas that is a successor of the affected version / bug fix mentioned in the respective issue. This may not be the issue. The other issue posted didn't seem to be the issue here either: self.homes_to_import_unnormalized_dd_df["district"] should not be a df object. I think I have tried materializing it to a pandas series (.compute()). If I recall, this raises the same or similar error.

Further, I did some testing to simulate the situation and perform the same operation: Using a comparable dataframe (same datatype columns), within a comparable mock - up of this utils class where the dask dataframe being passed in was created in the same scope (under if __name__ == "__main__" in the main script), but I created the dataframe through a different mechanism: python dict -> pd.DataFrame() -> dd.from_pandas().  The same apply function on this mock up of the same column worked as should be expected (no exception raised, a valid dataframe with 4 int columns returned as the result). 

With this, I'm led to believe that there is probably is a defect in the dataframe created by dd.read_sql_table() (from a view in Oracle) in this specific context that is adversely affecting downstream operations. This is why I posted the sequential history of upstream operations in the admittedly verbose initial description of the issue. I will traverse the code in the repo in .read_sql_table() and the functions it is calling to see if I can find an explanation for this failure (and hopefully a workaround) to get this one step in my pipeline out of a rut.

For the second issue, if anyone has any advice for ways I can introspect the offending dataframe and series objects involved, please let me know.  I am trying to look at any relevant attributes to see if there is any obvious defect that may explain this failure: Is a required attribute on this series object (or the dataframe it was subsetted from) missing or corrupted such that it would explain why "name" attribute not being found on the dataframe that is being created by the series object's .apply() method (several iterations into the .apply task). Also, is there any way to do this as the apply method is running? If I add to the function being applied (above the code being executed), a print statement: print([series element being worked with]) ..., It appears that several elements in the series are successfully processed. I see the result of the print statement a few dozen times with no exception raised, then an element in the series that appears to be valid will be processed. This iteration raises the AttributeError being described here... The fact that the series element that is breaking this .apply call seems to have no defect in it rules out a problem with the DATA in the series itself (e.g. np.nan() values, null entries, objects of a type not matching the type of the other elements of the series ...). Please let me know. For most objects in Python, I would call inspect.germembers() on it to look at its attributes and otherwise introspect it, but this doesn't seem to work here. It raises NotImplementedError. Sorry for the length and I hope this makes sense. If necessary I can work on the wording of this reply later when I have time.

**Comment by bryanwweber** [comment]
Hi @david-thrower Thanks again for the reply. If you can provide a minimal example to reproduce this issue, that would be very helpful, say 50 lines or less ideally, and also ideally not in a module/package structure. It seems like you did some testing in a smaller codebase to reproduce the problem; can you reduce that code even more? If the failure relies on having a SQL table, it'd be super handy if you could provide a small script to generate an appropriate table via the built-in `sqlite` module, that will make testing this issue much easier.

I'm asking for all this because as the issue stands now, I don't know what code I need to put into my editor to actually see the same error as you're getting. And without being able to put some code in my editor, it's really really hard to try and fix whatever is happening!

Last, if there are two separate problems, it would be helpful to focus this issue on one of the problems and open another issue for the other one, with a corresponding minimal reproducing code for that particular error. Thank you!

**Comment by david-thrower** [comment]
## Here is my speculative conclusion (perhaps confusion may be a better way to word it. I certainly don't claim to know everything here.):

Hopefully I am wording this well enough: It appears to me that perhaps [index metadata | indexing functionality] regarding list and list - like iterables may be getting [lost | corrupted] between the global scope and the subprocesses which I presume that .apply() may be using to execute the function in parallel over multiple partitions. More specifically, the issue appears to occur when the affected iterable was either:
1. Created within the applied function as a result of operations that were performed on the argument being passed in by .apply(). (The example I just posted).
2. AND / OR as I saw in another test which I can add when I have time to clean in up: The affected iterable itself is an object from the global scope being used within the scope of the function being applied and executed in the applied function.

If this is the case, If it does not already exist, we may want to consider adding a parameter to .apply() (and perhaps other parallel operations) that controls whether or not globals will be copied into the subprocesses being run (and if so which ones). For example, a param: copy_globals: Default - False (nothing from the global scope is copied other than what is generally required by Dask), True (all global objects copied into the subprocesses running - obviously rarely advisable to be used), or a user defined List or dict object listing the desired globals to be copied (like the exec() function has as the parameters globals, locals).

**Comment by jsignell** [comment]
Ah ok I know what the issue is! Thank you for taking the time to write it up so completely.

If you look at the full traceback you'll see that it starts with:

```python-traceback
ValueError: Metadata inference failed in `apply`.

You have supplied a custom function and Dask is unable to 
determine the type of output that that function returns. 

To resolve this please provide a meta= keyword.
The docstring of the Dask function you ran should have more information.

Original error is below:
------------------------
IndexError('index -2 is out of bounds for axis 0 with size 1')
```

I know that message can be hard to parse, but what it is saying is that the tiny version of your data that dask uses to keep track of what the output should look like (`meta`) is not good enough for this custom function. So it is recommending that you pass `meta` explicitly:

```python
# take a small subset of your data
sample = refractory_df_dd["district"].head()
# apply your function
meta = sample.apply(split_and_return)
# pass that in as the `meta`
refractory_result = refractory_df_dd["district"].apply(split_and_return, meta=meta)

**Comment by david-thrower** [comment]
@jsignell Thank you for that pointer. It did successfully correct the issue on my test script. I am surprised that this worked because I had already made a few unsuccessful attempts to manually parse the meta object. I had tried manually calling what if I remember right was `meta = dd.make_meta(...)` with the correct specification, I tried `meta = pd.DataFrame({'city':[],'state':[], 'zip_code':[]})`... Sorry to take up as much of your and the community's time.
